File: functions/go-func/proto2adif.go
package forester

import (
	"bytes"
	"fmt"
	"github.com/Matir/adifparser"
	adifpb "github.com/k0swe/adif-json-protobuf/go"
	"math"
	"strconv"
	"strings"
	"time"
)

func protoToAdif(adif *adifpb.Adif) (string, error) {
	buf := new(bytes.Buffer)
	writer := adifparser.NewADIFWriter(buf)
	for _, qso := range adif.Qsos {
		record := writeQso(qso)
		err := writer.WriteRecord(record)
		if err != nil {
			return "", err
		}
	}
	if err := writer.Flush(); err != nil {
		return "", err
	}
	return buf.String(), nil
}

func writeQso(qso *adifpb.Qso) adifparser.ADIFRecord {
	var rec = adifparser.NewADIFRecord()
	writeTopLevel(qso, rec)
	writeAppDefined(qso, rec)
	writeContactedStation(qso, rec)
	writeLoggingStation(qso, rec)
	writeContest(qso, rec)
	writePropagation(qso, rec)
	writeAwardsAndCredit(qso, rec)
	writeUploads(qso, rec)
	writeQsls(qso, rec)
	return rec
}

func writeTopLevel(qso *adifpb.Qso, rec adifparser.ADIFRecord) {
	writeString(rec, "band", qso.Band)
	writeString(rec, "band_rx", qso.BandRx)
	writeString(rec, "comment", qso.Comment)
	writeInt(rec, "distance", int64(qso.DistanceKm))
	writeFloat(rec, "freq", qso.Freq, 6)
	writeFloat(rec, "freq_rx", qso.FreqRx, 6)
	writeString(rec, "mode", qso.Mode)
	writeString(rec, "notes", qso.Notes)
	writeString(rec, "public_key", qso.PublicKey)
	writeString(rec, "qso_complete", qso.Complete)
	writeDate(rec, "qso_date", qso.TimeOn.AsTime())
	writeTime(rec, "time_on", qso.TimeOn.AsTime())
	writeDate(rec, "qso_date_off", qso.TimeOff.AsTime())
	writeTime(rec, "time_off", qso.TimeOff.AsTime())
	writeBool(rec, "qso_random", qso.Random)
	writeString(rec, "rst_rcvd", qso.RstReceived)
	writeString(rec, "rst_sent", qso.RstSent)
	writeString(rec, "submode", qso.Submode)
	writeBool(rec, "swl", qso.Swl)
}

func writeAppDefined(qso *adifpb.Qso, rec adifparser.ADIFRecord) {
	for k, v := range qso.AppDefined {
		rec.SetValue(k, v)
	}
}

func writeContactedStation(qso *adifpb.Qso, rec adifparser.ADIFRecord) {
	if qso.ContactedStation == nil {
		return
	}
	writeString(rec, "address", qso.ContactedStation.Address)
	writeInt(rec, "age", int64(qso.ContactedStation.Age))
	writeString(rec, "call", qso.ContactedStation.StationCall)
	writeString(rec, "cnty", qso.ContactedStation.County)
	writeString(rec, "cont", qso.ContactedStation.Continent)
	writeString(rec, "contacted_op", qso.ContactedStation.OpCall)
	writeString(rec, "country", qso.ContactedStation.Country)
	writeInt(rec, "cqz", int64(qso.ContactedStation.CqZone))
	writeString(rec, "darc_dok", qso.ContactedStation.DarcDok)
	writeInt(rec, "dxcc", int64(qso.ContactedStation.Dxcc))
	writeString(rec, "email", qso.ContactedStation.Email)
	writeString(rec, "eq_call", qso.ContactedStation.OwnerCall)
	writeInt(rec, "fists", int64(qso.ContactedStation.Fists))
	writeInt(rec, "fists_cc", int64(qso.ContactedStation.FistsCc))
	writeString(rec, "gridsquare", qso.ContactedStation.GridSquare)
	writeString(rec, "iota", qso.ContactedStation.Iota)
	writeInt(rec, "iota_island_id", int64(qso.ContactedStation.IotaIslandId))
	writeInt(rec, "ituz", int64(qso.ContactedStation.ItuZone))
	writeLatLon(rec, "lat", qso.ContactedStation.Latitude, true)
	writeLatLon(rec, "lon", qso.ContactedStation.Longitude, false)
	writeString(rec, "name", qso.ContactedStation.OpName)
	writeString(rec, "pfx", qso.ContactedStation.Pfx)
	writeString(rec, "qsl_via", qso.ContactedStation.QslVia)
	writeString(rec, "qth", qso.ContactedStation.City)
	writeString(rec, "region", qso.ContactedStation.Region)
	writeString(rec, "rig", qso.ContactedStation.Rig)
	writeFloat(rec, "rx_pwr", qso.ContactedStation.Power, 2)
	writeString(rec, "sig", qso.ContactedStation.Sig)
	writeString(rec, "sig_info", qso.ContactedStation.SigInfo)
	writeBool(rec, "silent_key", qso.ContactedStation.SilentKey)
	writeString(rec, "skcc", qso.ContactedStation.Skcc)
	writeString(rec, "sota_ref", qso.ContactedStation.SotaRef)
	writeString(rec, "state", qso.ContactedStation.State)
	writeInt(rec, "ten_ten", int64(qso.ContactedStation.TenTen))
	writeInt(rec, "uksmg", int64(qso.ContactedStation.Uksmg))
	writeString(rec, "usaca_counties", qso.ContactedStation.UsacaCounties)
	writeString(rec, "vucc_grids", qso.ContactedStation.VuccGrids)
	writeString(rec, "web", qso.ContactedStation.Web)
}

func writeLoggingStation(qso *adifpb.Qso, rec adifparser.ADIFRecord) {
	if qso.LoggingStation == nil {
		return
	}
	writeInt(rec, "ant_az", int64(qso.LoggingStation.AntennaAzimuth))
	writeInt(rec, "ant_el", int64(qso.LoggingStation.AntennaElevation))
	writeString(rec, "my_antenna", qso.LoggingStation.Antenna)
	writeString(rec, "my_city", qso.LoggingStation.City)
	writeString(rec, "my_cnty", qso.LoggingStation.County)
	writeString(rec, "my_country", qso.LoggingStation.Country)
	writeInt(rec, "my_cq_zone", int64(qso.LoggingStation.CqZone))
	writeInt(rec, "my_dxcc", int64(qso.LoggingStation.Dxcc))
	writeInt(rec, "my_fists", int64(qso.LoggingStation.Fists))
	writeString(rec, "my_gridsquare", qso.LoggingStation.GridSquare)
	writeString(rec, "my_iota", qso.LoggingStation.Iota)
	writeInt(rec, "my_iota_island_id", int64(qso.LoggingStation.IotaIslandId))
	writeInt(rec, "my_itu_zone", int64(qso.LoggingStation.ItuZone))
	writeLatLon(rec, "my_lat", qso.LoggingStation.Latitude, true)
	writeLatLon(rec, "my_lon", qso.LoggingStation.Longitude, false)
	writeString(rec, "my_name", qso.LoggingStation.OpName)
	writeString(rec, "my_postal_code", qso.LoggingStation.PostalCode)
	writeString(rec, "my_rig", qso.LoggingStation.Rig)
	writeString(rec, "my_sig", qso.LoggingStation.Sig)
	writeString(rec, "my_sig_info", qso.LoggingStation.SigInfo)
	writeString(rec, "my_sota_ref", qso.LoggingStation.SotaRef)
	writeString(rec, "my_state", qso.LoggingStation.State)
	writeString(rec, "my_street", qso.LoggingStation.Street)
	writeString(rec, "my_usaca_counties", qso.LoggingStation.UsacaCounties)
	writeString(rec, "my_vucc_grids", qso.LoggingStation.VuccGrids)
	writeString(rec, "operator", qso.LoggingStation.OpCall)
	writeString(rec, "owner_callsign", qso.LoggingStation.OwnerCall)
	writeString(rec, "station_callsign", qso.LoggingStation.StationCall)
	writeFloat(rec, "tx_pwr", qso.LoggingStation.Power, 2)
}

func writeContest(qso *adifpb.Qso, rec adifparser.ADIFRecord) {
	if qso.Contest == nil {
		return
	}
	writeString(rec, "contest_id", qso.Contest.ContestId)
	writeString(rec, "arrl_sect", qso.Contest.ArrlSection)
	writeString(rec, "class", qso.Contest.StationClass)
	writeString(rec, "check", qso.Contest.Check)
	writeString(rec, "precedence", qso.Contest.Precedence)
	writeString(rec, "srx", qso.Contest.SerialReceived)
	writeString(rec, "srx_string", qso.Contest.SerialReceived)
	writeString(rec, "stx", qso.Contest.SerialSent)
	writeString(rec, "stx_string", qso.Contest.SerialSent)
}

func writePropagation(qso *adifpb.Qso, rec adifparser.ADIFRecord) {
	if qso.Propagation == nil {
		return
	}
	writeInt(rec, "a_index", int64(qso.Propagation.AIndex))
	writeString(rec, "ant_path", qso.Propagation.AntPath)
	writeBool(rec, "force_init", qso.Propagation.ForceInit)
	writeInt(rec, "k_index", int64(qso.Propagation.KIndex))
	writeInt(rec, "max_bursts", int64(qso.Propagation.MaxBursts))
	writeString(rec, "ms_shower", qso.Propagation.MeteorShowerName)
	writeInt(rec, "nr_bursts", int64(qso.Propagation.NrBursts))
	writeInt(rec, "nr_pings", int64(qso.Propagation.NrPings))
	writeString(rec, "prop_mode", qso.Propagation.PropagationMode)
	writeString(rec, "sat_mode", qso.Propagation.SatMode)
	writeString(rec, "sat_name", qso.Propagation.SatName)
	writeInt(rec, "sfi", int64(qso.Propagation.SolarFluxIndex))
}

func writeAwardsAndCredit(qso *adifpb.Qso, rec adifparser.ADIFRecord) {
	writeAwards(qso.AwardSubmitted, rec, "award_submitted")
	writeAwards(qso.AwardGranted, rec, "award_granted")
	writeCredit(qso.CreditSubmitted, rec, "credit_submitted")
	writeCredit(qso.CreditGranted, rec, "credit_granted")
}

func writeAwards(awards []string, rec adifparser.ADIFRecord, field string) {
	writeString(rec, field, strings.Join(awards, ","))
}

func writeCredit(credit []*adifpb.Credit, rec adifparser.ADIFRecord, field string) {
	var elem = make([]string, 0)
	for _, c := range credit {
		e := c.Credit
		if c.QslMedium != "" {
			e += ":" + c.QslMedium
		}
		elem = append(elem, e)
	}
	writeString(rec, field, strings.Join(elem, ","))
}

func writeUploads(qso *adifpb.Qso, rec adifparser.ADIFRecord) {
	if qso.Qrzcom != nil {
		writeString(rec, "qrzcom_qso_upload_status", writeUploadStatus(qso.Qrzcom.UploadStatus))
		writeDate(rec, "qrzcom_qso_upload_date", qso.Qrzcom.UploadDate.AsTime())
	}

	if qso.Hrdlog != nil {
		writeString(rec, "hrdlog_qso_upload_status", writeUploadStatus(qso.Hrdlog.UploadStatus))
		writeDate(rec, "hrdlog_qso_upload_date", qso.Hrdlog.UploadDate.AsTime())
	}

	if qso.Clublog != nil {
		writeString(rec, "clublog_qso_upload_status", writeUploadStatus(qso.Clublog.UploadStatus))
		writeDate(rec, "clublog_qso_upload_date", qso.Clublog.UploadDate.AsTime())
	}
}

func writeUploadStatus(status adifpb.UploadStatus) string {
	switch status {
	case adifpb.UploadStatus_UPLOAD_COMPLETE:
		return "Y"
	case adifpb.UploadStatus_DO_NOT_UPLOAD:
		return "N"
	case adifpb.UploadStatus_MODIFIED_AFTER_UPLOAD:
		return "M"
	default:
		return ""
	}
}

func writeQsls(qso *adifpb.Qso, rec adifparser.ADIFRecord) {
	writeCardQsl(qso.Card, rec)
	writeQsl(qso.Eqsl, rec, "eqsl_")
	writeQsl(qso.Lotw, rec, "lotw_")
}

func writeCardQsl(qsl *adifpb.Qsl, rec adifparser.ADIFRecord) {
	if qsl == nil {
		return
	}
	writeQsl(qsl, rec, "")
	writeString(rec, "qsl_sent_via", qsl.SentVia)
	writeString(rec, "qsl_rcvd_via", qsl.ReceivedVia)
	writeString(rec, "qslmsg", qsl.ReceivedMessage)
}

func writeQsl(qsl *adifpb.Qsl, rec adifparser.ADIFRecord, prefix string) {
	if qsl == nil {
		return
	}
	writeString(rec, prefix+"qsl_sent", qsl.SentStatus)
	writeDate(rec, prefix+"qslsdate", qsl.SentDate.AsTime())
	writeString(rec, prefix+"qsl_rcvd", qsl.ReceivedStatus)
	writeDate(rec, prefix+"qslrdate", qsl.ReceivedDate.AsTime())

}

func writeString(rec adifparser.ADIFRecord, adifField string, value string) {
	if value == "" {
		return
	}
	rec.SetValue(adifField, value)
}

func writeBool(rec adifparser.ADIFRecord, adifField string, value bool) {
	if !value {
		return
	}
	rec.SetValue(adifField, "Y")
}

func writeInt(rec adifparser.ADIFRecord, adifField string, value int64) {
	if value == 0 {
		return
	}
	rec.SetValue(adifField, strconv.FormatInt(value, 10))
}

func writeFloat(rec adifparser.ADIFRecord, adifField string, value float64, precision int) {
	if value == 0 {
		return
	}
	rec.SetValue(adifField, strconv.FormatFloat(value, 'g', precision, 64))
}

func writeLatLon(rec adifparser.ADIFRecord, adifField string, latLon float64, isLat bool) {
	if latLon == 0 {
		return
	}
	value := latLonToString(latLon, isLat)
	rec.SetValue(adifField, value)
}

func latLonToString(latLon float64, isLat bool) string {
	var cardinal string
	if isLat {
		if latLon >= 0 {
			cardinal = "N"
		} else {
			cardinal = "S"
		}
	} else {
		if latLon >= 0 {
			cardinal = "E"
		} else {
			cardinal = "W"
		}
	}
	degrees := math.Floor(math.Abs(latLon))
	minutes := (math.Abs(latLon) - math.Abs(degrees)) * 60
	return fmt.Sprintf("%s%03d %06.3f", cardinal, int(degrees), minutes)
}

func writeDate(rec adifparser.ADIFRecord, adifField string, value time.Time) {
	if value.Unix() == 0 {
		return
	}
	// YYYYMMDD
	rec.SetValue(adifField, dateToString(value))
}

func dateToString(value time.Time) string {
	return value.Format("20060102")
}

func writeTime(rec adifparser.ADIFRecord, adifField string, value time.Time) {
	if value.Unix() == 0 {
		return
	}
	// HHMMSS
	rec.SetValue(adifField, timeToString(value))
}

func timeToString(value time.Time) string {
	return value.Format("150405")
}

File: web-old/src/app/shared/proto2adif.ts
import { SimpleAdif } from 'adif-parser-ts';

import {
  ContestData,
  Credit,
  Propagation,
  Qsl,
  Qso,
  Station,
  Upload,
  UploadStatus,
} from '../qso';

export class Proto2Adif {
  /**
   * Translate QSOs from Forester's internal format to AdifParser's SimpleAdif.
   */
  public static translateAdi(qsos: Qso[]): SimpleAdif {
    let myCall = 'N0CALL';
    if (qsos && qsos[0] && qsos[0].loggingStation) {
      myCall = qsos[0].loggingStation.stationCall;
    }
    const adif: SimpleAdif = {
      header: {
        text: `Log for ${myCall}`,
        adif_ver: '3.1.1',
        programid: 'Forester',
        programversion: `${new Date().toISOString()}`,
      },
      records: [],
    };
    qsos = qsos.sort((a, b) => a.timeOn.getTime() - b.timeOn.getTime());
    for (const qsoObj of qsos) {
      const qso = this.translateQso(qsoObj);
      adif.records.push(qso);
    }
    // get rid of `undefined`s
    return JSON.parse(JSON.stringify(adif));
  }

  private static translateQso(qso: Qso): { [p: string]: string } {
    const record: { [p: string]: string } = {};
    this.translateTopLevel(qso, record);
    this.translateAppDefined(qso, record);
    this.translateLoggingStation(qso.loggingStation, record);
    this.translateContactedStation(qso.contactedStation, record);
    this.translateContest(qso.contest, record);
    this.translatePropagation(qso.propagation, record);
    this.translateCreditAndAwards(qso, record);
    this.translateUploads(qso, record);
    this.translateQsls(qso, record);
    return record;
  }

  private static translateTopLevel(
    qso: Qso,
    record: { [p: string]: string },
  ): void {
    record.band = this.getString(qso.band);
    record.band_rx = this.getString(qso.bandRx);
    record.comment = this.getString(qso.comment);
    record.distance = this.getNumber(qso.distanceKm);
    record.freq = this.getNumber(qso.freq);
    record.freq_rx = this.getNumber(qso.freqRx);
    record.mode = this.getString(qso.mode);
    record.notes = this.getString(qso.notes);
    record.public_key = this.getString(qso.publicKey);
    record.qso_complete = this.getString(qso.complete);
    record.qso_date = this.getDate(qso.timeOn);
    record.time_on = this.getTime(qso.timeOn);
    record.qso_date_off = this.getDate(qso.timeOff);
    record.time_off = this.getTime(qso.timeOff);
    record.qso_random = this.getBool(qso.random);
    record.rst_rcvd = this.getString(qso.rstReceived);
    record.rst_sent = this.getString(qso.rstSent);
    record.submode = this.getString(qso.submode);
    record.swl = this.getBool(qso.swl);
  }

  private static translateAppDefined(
    qso: Qso,
    record: { [p: string]: string },
  ): void {
    if (!qso.appDefined) {
      return;
    }
    for (const key of Object.keys(qso.appDefined)) {
      record[key] = qso.appDefined[key];
    }
  }

  private static translateLoggingStation(
    loggingStation: Station,
    record: { [p: string]: string },
  ): void {
    record.ant_az = this.getNumber(loggingStation.antennaAzimuth);
    record.ant_el = this.getNumber(loggingStation.antennaElevation);
    record.my_antenna = this.getString(loggingStation.antenna);
    record.my_city = this.getString(loggingStation.city);
    record.my_cnty = this.getString(loggingStation.county);
    record.my_country = this.getString(loggingStation.country);
    record.my_cq_zone = this.getNumber(loggingStation.cqZone);
    record.my_dxcc = this.getNumber(loggingStation.dxcc);
    record.my_fists = this.getNumber(loggingStation.fists);
    record.my_gridsquare = this.getString(loggingStation.gridSquare);
    record.my_iota = this.getString(loggingStation.iota);
    record.my_iota_island_id = this.getNumber(loggingStation.iotaIslandId);
    record.my_itu_zone = this.getNumber(loggingStation.ituZone);
    record.my_lat = this.getLatLon(loggingStation.latitude, true);
    record.my_lon = this.getLatLon(loggingStation.longitude, false);
    record.my_name = this.getString(loggingStation.opName);
    record.my_postal_code = this.getString(loggingStation.postalCode);
    record.my_rig = this.getString(loggingStation.rig);
    record.my_sig = this.getString(loggingStation.sig);
    record.my_sig_info = this.getString(loggingStation.sigInfo);
    record.my_sota_ref = this.getString(loggingStation.sotaRef);
    record.my_state = this.getString(loggingStation.state);
    record.my_street = this.getString(loggingStation.street);
    record.my_usaca_counties = this.getString(loggingStation.usacaCounties);
    record.my_vucc_grids = this.getString(loggingStation.vuccGrids);
    record.operator = this.getString(loggingStation.opCall);
    record.owner_callsign = this.getString(loggingStation.ownerCall);
    record.station_callsign = this.getString(loggingStation.stationCall);
    record.tx_pwr = this.getNumber(loggingStation.power);
  }

  private static translateContactedStation(
    contactedStation: Station,
    record: { [p: string]: string },
  ): void {
    record.address = contactedStation.address;
    record.age = this.getNumber(contactedStation.age);
    record.call = this.getString(contactedStation.stationCall);
    record.cnty = this.getString(contactedStation.county);
    record.cont = this.getString(contactedStation.continent);
    record.contacted_op = this.getString(contactedStation.opCall);
    record.country = this.getString(contactedStation.country);
    record.cqz = this.getNumber(contactedStation.cqZone);
    record.darc_dok = this.getString(contactedStation.darcDok);
    record.dxcc = this.getNumber(contactedStation.dxcc);
    record.email = this.getString(contactedStation.email);
    record.eq_call = this.getString(contactedStation.ownerCall);
    record.fists = this.getNumber(contactedStation.fists);
    record.fists_cc = this.getNumber(contactedStation.fistsCc);
    record.gridsquare = this.getString(contactedStation.gridSquare);
    record.iota = this.getString(contactedStation.iota);
    record.iota_island_id = this.getNumber(contactedStation.iotaIslandId);
    record.ituz = this.getNumber(contactedStation.ituZone);
    record.lat = this.getLatLon(contactedStation.latitude, true);
    record.lon = this.getLatLon(contactedStation.longitude, false);
    record.name = this.getString(contactedStation.opName);
    record.pfx = this.getString(contactedStation.pfx);
    record.qsl_via = this.getString(contactedStation.qslVia);
    record.qth = this.getString(contactedStation.city);
    record.region = this.getString(contactedStation.region);
    record.rig = this.getString(contactedStation.rig);
    record.rx_pwr = this.getNumber(contactedStation.power);
    record.sig = this.getString(contactedStation.sig);
    record.sig_info = this.getString(contactedStation.sigInfo);
    record.silent_key = this.getBool(contactedStation.silentKey);
    record.skcc = this.getString(contactedStation.skcc);
    record.sota_ref = this.getString(contactedStation.sotaRef);
    record.state = this.getString(contactedStation.state);
    record.ten_ten = this.getNumber(contactedStation.tenTen);
    record.uksmg = this.getNumber(contactedStation.uksmg);
    record.usaca_counties = this.getString(contactedStation.usacaCounties);
    record.vucc_grids = this.getString(contactedStation.vuccGrids);
    record.web = this.getString(contactedStation.web);
  }

  private static translateContest(
    contest: ContestData,
    record: { [p: string]: string },
  ): void {
    if (!contest) {
      return;
    }
    record.contest_id = this.getString(contest.contestId);
    record.arrl_sect = this.getString(contest.arrlSection);
    record.class = this.getString(contest.stationClass);
    record.check = this.getString(contest.check);
    record.precedence = this.getString(contest.precedence);
    record.srx_string = this.getString(contest.serialReceived);
    record.stx_string = this.getString(contest.serialSent);
  }

  private static translatePropagation(
    prop: Propagation,
    record: { [p: string]: string },
  ): void {
    if (!prop) {
      return;
    }
    record.a_index = this.getNumber(prop.aIndex);
    record.ant_path = this.getString(prop.antPath);
    record.force_init = this.getBool(prop.forceInit);
    record.k_index = this.getNumber(prop.kIndex);
    record.max_bursts = this.getNumber(prop.maxBursts);
    record.ms_shower = this.getString(prop.meteorShowerName);
    record.nr_bursts = this.getNumber(prop.nrBursts);
    record.nr_pings = this.getNumber(prop.nrPings);
    record.prop_mode = this.getString(prop.propagationMode);
    record.sat_mode = this.getString(prop.satMode);
    record.sat_name = this.getString(prop.satName);
    record.sfi = this.getNumber(prop.solarFluxIndex);
  }

  private static translateCreditAndAwards(
    qso: Qso,
    record: { [p: string]: string },
  ): void {
    record.award_submitted = this.translateAwards(qso.awardSubmitted);
    record.award_granted = this.translateAwards(qso.awardGranted);
    record.credit_submitted = this.translateCredit(qso.creditSubmitted);
    record.credit_granted = this.translateCredit(qso.creditGranted);
  }

  private static translateAwards(awards: Array<string>): string {
    if (!awards) {
      return undefined;
    }
    return awards.join(',');
  }

  private static translateCredit(credits: Credit[]): string {
    if (!credits) {
      return undefined;
    }

    return Object.keys(credits)
      .map((k) => {
        const c = credits[k];
        if (c.qslMedium) {
          return `${c.credit}:${c.qslMedium}`;
        }
        return c.credit;
      })
      .join(',');
  }

  private static translateUploads(
    qso: Qso,
    record: { [p: string]: string },
  ): void {
    this.translateUpload(qso.qrzcom, 'qrzcom', record);
    this.translateUpload(qso.hrdlog, 'hrdlog', record);
    this.translateUpload(qso.clublog, 'clublog', record);
  }

  private static translateUpload(
    upload: Upload,
    uploadProvider: string,
    record: { [p: string]: string },
  ): void {
    if (!upload) {
      return;
    }
    let status: any = upload.uploadStatus;
    if (typeof status === 'string') {
      // This should probably happen in QsoService.unmarshal somewhere...
      status = UploadStatus[`${status}`];
    }
    switch (status) {
      case UploadStatus.UPLOAD_COMPLETE:
        record[uploadProvider + '_qso_upload_status'] = 'Y';
        break;
      case UploadStatus.DO_NOT_UPLOAD:
        record[uploadProvider + '_qso_upload_status'] = 'N';
        break;
      case UploadStatus.MODIFIED_AFTER_UPLOAD:
        record[uploadProvider + '_qso_upload_status'] = 'M';
        break;
    }
    record[uploadProvider + '_qso_upload_date'] = this.getDate(
      upload.uploadDate,
    );
  }

  private static translateQsls(
    qso: Qso,
    record: { [p: string]: string },
  ): void {
    this.translateCardQsl(qso.card, record);
    this.translateQsl(qso.eqsl, 'eqsl_', record);
    this.translateQsl(qso.lotw, 'lotw_', record);
  }

  private static translateCardQsl(
    qsl: Qsl,
    record: { [p: string]: string },
  ): void {
    if (qsl == null) {
      return undefined;
    }
    this.translateQsl(qsl, '', record);
    record.qsl_sent_via = this.getString(qsl.sentVia);
    record.qsl_rcvd_via = this.getString(qsl.receivedVia);
    record.qslmsg = this.getString(qsl.receivedMessage);
  }

  private static translateQsl(
    qsl: Qsl,
    qslProvider: string,
    record: { [p: string]: string },
  ): void {
    if (!qsl) {
      return;
    }
    record[qslProvider + 'qsl_sent'] = this.getString(qsl.sentStatus);
    record[qslProvider + 'qsl_rcvd'] = this.getString(qsl.receivedStatus);
    record[qslProvider + 'qslsdate'] = this.getDate(qsl.sentDate);
    record[qslProvider + 'qslrdate'] = this.getDate(qsl.receivedDate);
  }

  private static getString(str: string): string | undefined {
    return !!str ? str : undefined;
  }

  private static getNumber(num: number): string | undefined {
    return num ? num.toString(10) : undefined;
  }

  private static getBool(bool: boolean): string | undefined {
    if (bool == null) {
      return undefined;
    }
    return bool ? 'Y' : 'N';
  }

  private static getLatLon(coord: number, isLat: boolean): string {
    if (!coord || coord === 0) {
      return undefined;
    }
    let cardinal;
    if (isLat) {
      if (coord > 0) {
        cardinal = 'N';
      } else {
        cardinal = 'S';
      }
    } else {
      if (coord > 0) {
        cardinal = 'E';
      } else {
        cardinal = 'W';
      }
    }
    const myCoord = Math.abs(coord);
    const degrees = Math.floor(myCoord);
    const degStr = degrees.toString().padStart(3, '0');
    const minutes = ((myCoord - degrees) * 60).toFixed(3).padStart(6, '0');
    return `${cardinal}${degStr} ${minutes}`;
  }

  private static getDate(date: Date): string {
    if (!date) {
      return undefined;
    }
    if (!(date instanceof Date)) {
      // This is kind of a band-aid. Really, dates should be fixed in
      // QsoService.unmarshalDates()
      date = new Date(date);
    }
    if (isNaN(date.getTime())) {
      // Invalid date
      return undefined;
    }
    const year = date.getUTCFullYear().toString();
    const month = (date.getUTCMonth() + 1).toString().padStart(2, '0');
    const day = date.getUTCDate().toString().padStart(2, '0');
    return `${year}${month}${day}`;
  }

  private static getTime(date: Date): string {
    if (!date) {
      return undefined;
    }
    const hour = date.getUTCHours().toString().padStart(2, '0');
    const minute = date.getUTCMinutes().toString().padStart(2, '0');
    const second = date.getUTCSeconds().toString().padStart(2, '0');
    return `${hour}${minute}${second}`;
  }

  private constructor() {}
}
